note/note_rfc6962.go

```diff
@@ -0,0 +1,239 @@
+	sigBytes = append(sigBytes, sth.TreeHeadSignature...)
+	sigLine := fmt.Sprintf("\u2014 %s %s", logName, base64.StdEncoding.EncodeToString(sigBytes))
+
+	return []byte(fmt.Sprintf("%s\n%s\n", body, sigLine)), nil
+}
+
+func rfc6962Keyhash(name string, logID [32]byte) uint32 {
+	h := sha256.New()
+	h.Write([]byte(name))
+	h.Write([]byte{0x0A, algRFC6962STH})
+	h.Write(logID[:])
+	r := h.Sum(nil)
+	return binary.BigEndian.Uint32(r)
+}
+
+// rfc6962LogName returns a sunlight checkpoint compatible log name from the
+// passed in CT log root URL.
+//
+// "For example, a log with submission prefix https://rome.ct.example.com/2024h1/ will use rome.ct.example.com/2024h1 as the checkpoint origin line"
+func rfc6962LogName(logURL string) string {
+	logURL = strings.ToLower(logURL)
+	logURL = strings.TrimPrefix(logURL, "http://")
+	logURL = strings.TrimPrefix(logURL, "https://")
+	logURL = strings.TrimSuffix(logURL, "/")
+	return logURL
+}
+
+type rfc6962Verifer struct {
+	name    string
+	keyHash uint32
+	v       func(msg []byte, origin string, sig []byte) bool
+}
+
+// Name returns the name associated with the key this verifier is based on.
+func (v *rfc6962Verifer) Name() string {
+	return v.name
+}
+
+// KeyHash returns a truncated hash of the key this verifier is based on.
+func (v *rfc6962Verifer) KeyHash() uint32 {
+	return v.keyHash
+}
+
+// Verify checks that the provided sig is valid over msg for the key this verifier is based on.
+func (v *rfc6962Verifer) Verify(msg, sig []byte) bool {
+	return v.v(msg, v.name, sig)
+}
+
+func verifyRFC6962(key crypto.PublicKey) func([]byte, string, []byte) bool {
+	return func(msg []byte, origin string, sig []byte) bool {
+		if len(sig) < timestampSize {
+			return false
+		}
+		t := binary.BigEndian.Uint64(sig)
+		sig = sig[timestampSize:]
+		hAlg := sig[0]
+		sAlg := sig[1]
+		sig = sig[2:]
+		sigLen := binary.BigEndian.Uint16(sig)
+		sig = sig[2:]
+		if len(sig) != int(sigLen) {
+			return false
+		}
+
+		// SHA256 (RFC 5246 s7.4.1.4.1.)
+		if hAlg != 0x04 {
+			return false
+		}
+
+		o, m, err := formatRFC6962STH(t, msg)
+		if err != nil {
+			return false
+		}
+		if origin != o {
+			return false
+		}
+		dgst := sha256.Sum256(m)
+		switch k := key.(type) {
+		case *ecdsa.PublicKey:
+			// RFC 5246 s7.4.1.4.1.
+			if sAlg != 0x03 {
+				return false
+			}
+			return ecdsa.VerifyASN1(k, dgst[:], sig)
+		case *rsa.PublicKey:
+			// RFC 5246 s7.4.1.4.1.
+			if sAlg != 0x01 {
+				return false
+			}
+			return rsa.VerifyPKCS1v15(k, crypto.SHA256, dgst[:], sig) != nil
+		default:
+			return false
+		}
+	}
+}
+
+// formatRFC6962STH uses the provided timestamp and checkpoint body to
+// recreate the RFC6962 STH structure over which the signature was made.
+func formatRFC6962STH(t uint64, msg []byte) (string, []byte, error) {
+	// Must be:
+	// origin (schema-less log root url)
+	// tree size (decimal)
+	// root hash (b64)
+	lines := strings.Split(string(msg), "\n")
+	if len(lines) != 4 {
```

Comment:
nit: empty line at the end? since you have a "Must be" above, might as well put it there

Reply:
I've added explicit "\n" on the comment, does that help?